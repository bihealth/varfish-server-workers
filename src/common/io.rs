//! Common, IO-related code.

use std::{
    fs::File,
    io::{BufRead, BufReader, BufWriter, Write},
    path::Path,
};

use flate2::{bufread::MultiGzDecoder, write::GzEncoder, Compression};

/// Transparently open a file with gzip decoder.
pub fn open_read_maybe_gz<P>(path: P) -> Result<Box<dyn BufRead>, anyhow::Error>
where
    P: AsRef<Path>,
{
    if path.as_ref().extension().map(|s| s.to_str()) == Some(Some("gz")) {
        tracing::trace!("Opening {:?} as gzip for reading", path.as_ref());
        let file = File::open(path)?;
        let bufreader = BufReader::new(file);
        let decoder = MultiGzDecoder::new(bufreader);
        Ok(Box::new(BufReader::new(decoder)))
    } else {
        tracing::trace!("Opening {:?} as plain text for reading", path.as_ref());
        let file = File::open(path).map(BufReader::new)?;
        Ok(Box::new(BufReader::new(file)))
    }
}

/// Transparently open afile with gzip encoder.
pub fn open_write_maybe_gz<P>(path: P) -> Result<Box<dyn Write>, anyhow::Error>
where
    P: AsRef<Path>,
{
    if path.as_ref().extension().map(|s| s.to_str()) == Some(Some("gz")) {
        tracing::trace!("Opening {:?} as gzip for writing", path.as_ref());
        let file = File::create(path)?;
        let bufwriter = BufWriter::new(file);
        let encoder = GzEncoder::new(bufwriter, Compression::default());
        Ok(Box::new(encoder))
    } else {
        tracing::trace!("Opening {:?} as plain text for writing", path.as_ref());
        let file = File::create(path)?;
        Ok(Box::new(file))
    }
}

// The output is wrapped in a Result to allow matching on errors
// Returns an Iterator to the Reader of the lines of the file.
pub fn read_lines<P: AsRef<Path>>(
    filename: P,
) -> std::io::Result<std::io::Lines<std::io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(std::io::BufReader::new(file).lines())
}

#[cfg(test)]
mod test {
    use std::io::Read;

    #[rstest::rstest]
    #[case(true)]
    #[case(false)]
    fn open_write_maybe_gz(#[case] is_gzip: bool) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{:?}", is_gzip);

        let filename = if is_gzip { "test.txt" } else { "test.txt.gz" };
        let tmp_dir = temp_testdir::TempDir::default();

        {
            let mut f = super::open_write_maybe_gz(tmp_dir.join(filename))?;
            f.flush()?;
        }

        let mut f = std::fs::File::open(tmp_dir.join(filename)).map(std::io::BufReader::new)?;
        let mut buf = Vec::new();
        f.read_to_end(&mut buf)?;

        insta::assert_snapshot!(format!("{:x?}", &buf));

        Ok(())
    }

    #[test]
    fn read_lines() -> Result<(), anyhow::Error> {
        let lines = super::read_lines("tests/common/lines.txt")?.collect::<Result<Vec<_>, _>>()?;

        insta::assert_yaml_snapshot!(lines);

        Ok(())
    }
}
