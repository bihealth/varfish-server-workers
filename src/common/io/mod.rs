//! Common, IO-related code.

pub mod std;
