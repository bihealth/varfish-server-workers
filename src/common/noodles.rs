//! Helper code for using noodles with S3.
//!
//! At the moment, we only support reading but not yet writing.
//!
//! The main reason is that the awslabs SDK for Rust currently supports easily
//! getting an `AsyncRead` from an S3 object but not an `AsyncWrite`.  For
//! this, we would have to create a wrapper that writes to a multipart upload
//! or similar.

use async_compression::tokio::bufread::GzipDecoder;
use mehari::common::{
    io::{std::is_gz, tokio::open_read_maybe_gz},
    noodles::AsyncVcfReader,
};
use noodles_bgzf as bgzf;
use noodles_core::Position;
use noodles_csi::{self as csi, index::reference_sequence::bin::Chunk};
use noodles_tabix as tabix;
use noodles_vcf as vcf;
use std::{path::Path, pin::Pin};
use tokio::io::{AsyncBufRead, BufReader};

/// Build TBI for file at `path_src` and write to `path_dst`.
pub async fn build_tbi<S, D>(path_src: S, path_dst: D) -> Result<(), anyhow::Error>
where
    S: AsRef<std::path::Path>,
    D: AsRef<std::path::Path>,
{
    let mut reader = tokio::fs::File::open(path_src.as_ref())
        .await
        .map(bgzf::AsyncReader::new)
        .map(vcf::AsyncReader::new)
        .map_err(|e| anyhow::anyhow!("error input file for tbi creation: {}", e))?;

    let header = reader
        .read_header()
        .await
        .map_err(|e| anyhow::anyhow!("error reading header: {}", e))?;

    let mut record = vcf::Record::default();

    let mut indexer = tabix::index::Indexer::default();
    indexer.set_header(csi::index::header::Builder::vcf().build());

    let mut start_position = reader.get_ref().virtual_position();

    while reader
        .read_record(&header, &mut record)
        .await
        .map_err(|e| anyhow::anyhow!("problem reading record: {}", e))?
        != 0
    {
        let end_position = reader.get_ref().virtual_position();
        let chunk = Chunk::new(start_position, end_position);

        let reference_sequence_name = record.chromosome().to_string();
        let start = Position::try_from(usize::from(record.position()))
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
            .map_err(|e| anyhow::anyhow!("error converting start position: {}", e))?;
        let end = record
            .end()
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
            .and_then(|position| {
                Position::try_from(usize::from(position))
                    .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
            })
            .map_err(|e| anyhow::anyhow!("error converting end position: {}", e))?;

        indexer
            .add_record(&reference_sequence_name, start, end, chunk)
            .map_err(|e| anyhow::anyhow!("error adding record to tabix index: {}", e))?;

        start_position = end_position;
    }

    let index = indexer.build();

    {
        let mut writer = tokio::fs::File::create(path_dst.as_ref())
            .await
            .map(tokio::io::BufWriter::new)
            .map(tabix::AsyncWriter::new)
            .map_err(|e| anyhow::anyhow!("error output file for tbi creation: {}", e))?;
        writer
            .write_index(&index)
            .await
            .map_err(|e| anyhow::anyhow!("error writing tabix index: {}", e))?;
        writer
            .shutdown()
            .await
            .map_err(|e| anyhow::anyhow!("error flushing tabix index: {}", e))?;

        tokio::time::sleep(std::time::Duration::from_millis(200)).await;
    }

    Ok(())
}

/// Open plain text or gzip reader via S3.
pub async fn s3_open_read_maybe_gz<P>(path: P) -> Result<Pin<Box<dyn AsyncBufRead>>, anyhow::Error>
where
    P: AsRef<Path>,
{
    let path_string = format!("{}", path.as_ref().display());
    let (bucket, key) = if let Some((bucket, key)) = path_string.split_once('/') {
        (bucket.to_string(), key.to_string())
    } else {
        anyhow::bail!("invalid S3 path: {}", path.as_ref().display());
    };

    let config = aws_config::load_from_env().await;
    let client = aws_sdk_s3::Client::new(&config);
    let object = client.get_object().bucket(&bucket).key(&key).send().await?;

    let path_is_gzip = is_gz(path.as_ref());
    tracing::trace!(
        "Opening S3 object {} as {} for reading (async)",
        path.as_ref().display(),
        if path_is_gzip {
            "gzip (allow multi-member)"
        } else {
            "plain text"
        }
    );

    if path_is_gzip {
        let bufreader = BufReader::new(object.body.into_async_read());
        let decoder = {
            let mut decoder = GzipDecoder::new(bufreader);
            decoder.multiple_members(true);
            decoder
        };
        Ok(Box::pin(BufReader::new(decoder)))
    } else {
        Ok(Box::pin(BufReader::new(object.body.into_async_read())))
    }
}

/// Helper function that opens a list of paths as VCF readers.
pub async fn open_vcf_readers(paths: &[String]) -> Result<Vec<AsyncVcfReader>, anyhow::Error> {
    let mut result = Vec::new();
    for path in paths.iter() {
        let buf_read = if super::s3::s3_mode() && !path.starts_with('/') {
            s3_open_read_maybe_gz(path).await?
        } else {
            open_read_maybe_gz(path).await?
        };
        result.push(AsyncVcfReader::new(buf_read));
    }
    Ok(result)
}

/// Helper function that opens one VCF reader at the given path.
///
/// The behaviour is as follows:
///
/// - If `path_in` is "-" then open stdin and read as plain text.
/// - If environment variable `AWS_PROFILE` is set to "varfish-s3" then enable S3 mode.
/// - If `path_in` is absolute or S3 mode is disabled then open `path_in` as local file
/// - Otherwise, attempt to open `path_in` as S3 object.
pub async fn open_vcf_reader(path_in: &str) -> Result<AsyncVcfReader, anyhow::Error> {
    let s3_mode = match std::env::var("AWS_PROFILE") {
        Ok(s) => s == "varfish-s3",
        _ => false,
    };
    if s3_mode && !path_in.starts_with('/') {
        Ok(vcf::AsyncReader::new(
            s3_open_read_maybe_gz(path_in)
                .await
                .map_err(|e| anyhow::anyhow!("could not build VCF reader from S3 file: {}", e))?,
        ))
    } else {
        Ok(vcf::AsyncReader::new(
            open_read_maybe_gz(path_in).await.map_err(|e| {
                anyhow::anyhow!("could not build VCF reader from local file: {}", e)
            })?,
        ))
    }
}

#[cfg(test)]
mod test {
    #[tokio::test]
    async fn build_tbi() -> Result<(), anyhow::Error> {
        let tmpdir = temp_testdir::TempDir::default();

        let path_in: String = "tests/seqvars/ingest/NA12878_dragen.vcf.gz".into();
        let path_out: String = tmpdir
            .join("out.vcf.gz.tbi")
            .to_str()
            .expect("invalid path")
            .into();
        super::build_tbi(&path_in, &path_out).await?;

        // This must be fixed first https://github.com/zaeleus/noodles/issues/213
        // let mut buffer: Vec<u8> = Vec::new();
        // hxdmp::hexdump(&crate::common::read_to_bytes(&path_out)?, &mut buffer)?;
        // insta::assert_snapshot!(String::from_utf8_lossy(&buffer));

        Ok(())
    }
}
