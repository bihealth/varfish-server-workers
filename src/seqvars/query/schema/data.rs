//! Code for representing variants internally, corresponds to what is written
//! out by `seqvars ingest`.

use noodles::vcf;

use crate::common::genotype_to_string;

/// Trait for attempting conversion from VCF record.
pub trait TryFromVcf: Sized {
    /// Error type to use.
    type Error;

    /// Convert from VCF record.
    ///
    /// # Arguments
    ///
    /// * `record` - VCF record.
    /// * `header` - VCF header.
    ///
    /// # Errors
    ///
    /// Returns an error if the record cannot be converted.
    fn try_from_vcf(
        record: &vcf::variant::RecordBuf,
        header: &vcf::Header,
    ) -> Result<Self, Self::Error>;
}

/// Information on the call as written out by ingest.
///
/// Corresponds to `FORMAT/*` in VCF.
///
/// Note that the ingested files have exactly one alternate allele.
#[derive(Debug, Clone, Default, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct CallInfo {
    /// The sample name.
    pub sample: String,
    /// The genotype, if applicable, e.g., "0/1"
    pub genotype: Option<String>,
    /// Genotype quality score, if applicable
    pub gq: Option<f32>,
    /// Total read coverage at site in the sample.
    pub dp: Option<i32>,
    /// Alternate allele depth for the single allele in the sample.
    pub ad: Option<i32>,
    /// Physical phasing ID for this sample.
    pub ps: Option<i32>,
}

impl Eq for CallInfo {}

/// pub trait for total allele counts.
pub trait An {
    /// Number of covered alleles.
    fn an(&self) -> i32;
}

/// pub trait for variant alternate allele counts.
pub trait Ac {
    /// Number of homozygous/homoplasmic carriers.
    fn hom(&self) -> i32;
    /// Number of heterozygous/heteroplasmic carriers.
    fn het(&self) -> i32;
    /// Number of total alternate alleles.
    fn ac(&self) -> i32;
}

/// pub trait for allele frequency.
pub trait Af {
    /// Allele frequency.
    fn af(&self) -> f32;
}

/// Blanket implementation of `Af` for any type that implements `Ac` and `An`.
impl<T: Ac + An> Af for T {
    fn af(&self) -> f32 {
        if self.an() == 0 {
            0.0
        } else {
            self.ac() as f32 / self.an() as f32
        }
    }
}

/// pub trait for carrier count.
pub trait Carriers {
    /// Total number of carriers.
    fn carriers(&self) -> i32;
}

/// pub trait for nuclear variant alternate allele including hemizygous.
pub trait Hemi: Ac {
    /// Number of hemizygous carriers.
    fn hemi(&self) -> i32;
}

/// gnomAD Population frequencies for variants on nuclear chromosomes.
#[derive(Debug, Clone, Default, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct NuclearFrequencies {
    /// Number of alleles.
    pub an: i32,
    /// Number of homozygous carriers.
    pub hom: i32,
    /// Number of heterozygous carriers.
    pub het: i32,
    /// Number of hemizygous carriers.
    pub hemi: i32,
}

impl An for NuclearFrequencies {
    fn an(&self) -> i32 {
        self.an
    }
}

impl Ac for NuclearFrequencies {
    fn hom(&self) -> i32 {
        self.hom
    }
    fn het(&self) -> i32 {
        self.het
    }
    fn ac(&self) -> i32 {
        2 * self.hom + self.het + self.hemi
    }
}

impl Hemi for NuclearFrequencies {
    fn hemi(&self) -> i32 {
        self.hemi
    }
}

impl Carriers for NuclearFrequencies {
    fn carriers(&self) -> i32 {
        self.hom + self.het + self.hemi
    }
}

/// Population frequencies for variants on the mitochondrial chromosome.
#[derive(Debug, Clone, Default, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct MitochondrialFrequencies {
    /// Number of alleles.
    pub an: i32,
    /// Number of homoplasmic carriers.
    pub hom: i32,
    /// Number of heteroplasmic carriers.
    pub het: i32,
}

impl An for MitochondrialFrequencies {
    fn an(&self) -> i32 {
        self.an
    }
}

impl Ac for MitochondrialFrequencies {
    fn hom(&self) -> i32 {
        self.hom
    }
    fn het(&self) -> i32 {
        self.het
    }
    fn ac(&self) -> i32 {
        self.hom + self.het
    }
}

impl Carriers for MitochondrialFrequencies {
    fn carriers(&self) -> i32 {
        self.hom + self.het
    }
}

/// In-house frequencies.
#[derive(Debug, Clone, Default, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct InHouseFrequencies {
    // Number of alleles.
    pub an: i32,
    /// Number of homozygous carriers.
    pub hom: i32,
    /// Number of heterozygous carriers.
    pub het: i32,
    /// Number of hemizygous carriers.
    pub hemi: i32,
}

impl Ac for InHouseFrequencies {
    fn hom(&self) -> i32 {
        self.hom
    }
    fn het(&self) -> i32 {
        self.het
    }
    fn ac(&self) -> i32 {
        2 * self.hom + self.het + self.hemi
    }
}

impl An for InHouseFrequencies {
    fn an(&self) -> i32 {
        self.an
    }
}

impl Carriers for InHouseFrequencies {
    fn carriers(&self) -> i32 {
        self.hom + self.het + self.hemi
    }
}

/// Population frequencies for a variant.
#[derive(Debug, Clone, Default, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct PopulationFrequencies {
    /// Frequencies in gnomAD-exomes.
    pub gnomad_exomes: NuclearFrequencies,
    /// Frequencies in gnomAD-genomes.
    pub gnomad_genomes: NuclearFrequencies,
    /// Frequencies in gnomAD-MT.
    pub gnomad_mtdna: MitochondrialFrequencies,
    /// Frequencies in HelixMtDb.
    pub helixmtdb: MitochondrialFrequencies,
    /// Frequencies in In-house database.
    pub inhouse: InHouseFrequencies,
}

/// Supporting code for `PopulationFrequencies`.
pub(crate) mod population_frequencies {
    /// Error type for `FromVcf` implementation.
    #[derive(thiserror::Error, Debug, Clone)]
    pub enum Error {
        // infallibe
    }
}

impl TryFromVcf for PopulationFrequencies {
    type Error = population_frequencies::Error;

    fn try_from_vcf(
        record: &vcf::variant::RecordBuf,
        _header: &vcf::Header,
    ) -> Result<Self, Self::Error> {
        use vcf::variant::record_buf::info::field::Value;

        macro_rules! extract_key {
            ($key:ident) => {
                let $key =
                    if let Some(Some(Value::Integer($key))) = record.info().get(stringify!($key)) {
                        *$key
                    } else {
                        0
                    };
            };
        }

        extract_key!(gnomad_exomes_an);
        extract_key!(gnomad_exomes_hom);
        extract_key!(gnomad_exomes_het);
        extract_key!(gnomad_exomes_hemi);

        extract_key!(gnomad_genomes_an);
        extract_key!(gnomad_genomes_hom);
        extract_key!(gnomad_genomes_het);
        extract_key!(gnomad_genomes_hemi);

        extract_key!(gnomad_mtdna_an);
        extract_key!(gnomad_mtdna_hom);
        extract_key!(gnomad_mtdna_het);

        extract_key!(helix_an);
        extract_key!(helix_hom);
        extract_key!(helix_het);

        Ok(PopulationFrequencies {
            gnomad_exomes: NuclearFrequencies {
                an: gnomad_exomes_an,
                hom: gnomad_exomes_hom,
                het: gnomad_exomes_het,
                hemi: gnomad_exomes_hemi,
            },
            gnomad_genomes: NuclearFrequencies {
                an: gnomad_genomes_an,
                hom: gnomad_genomes_hom,
                het: gnomad_genomes_het,
                hemi: gnomad_genomes_hemi,
            },
            gnomad_mtdna: MitochondrialFrequencies {
                an: gnomad_mtdna_an,
                hom: gnomad_mtdna_hom,
                het: gnomad_mtdna_het,
            },
            helixmtdb: MitochondrialFrequencies {
                an: helix_an,
                hom: helix_hom,
                het: helix_het,
            },
            inhouse: Default::default(),
        })
    }
}

/// Sequence variant representation VCF-style.
#[derive(Debug, Clone, Default, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct VcfVariant {
    /// Chromosome.
    pub chrom: String,
    /// Position.
    pub pos: i32,
    /// Reference allele.
    pub ref_allele: String,
    /// Alternate allele.
    pub alt_allele: String,
}

impl From<VcfVariant> for annonars::common::spdi::Var {
    fn from(val: VcfVariant) -> Self {
        annonars::common::spdi::Var::new(
            annonars::common::cli::canonicalize(&val.chrom),
            val.pos,
            val.ref_allele,
            val.alt_allele,
        )
    }
}

/// Supporting code for `VcfVariant`.
pub(crate) mod vcf_variant {
    /// Error type for `FromVcf` implementation.
    #[derive(thiserror::Error, Debug, Clone)]
    pub enum Error {
        #[error("Missing POS value")]
        MissingVariantStart,
        #[error("Missing ALT values")]
        MissingAlternateBases,
    }
}

impl TryFromVcf for VcfVariant {
    type Error = vcf_variant::Error;

    fn try_from_vcf(
        record: &vcf::variant::RecordBuf,
        _header: &vcf::Header,
    ) -> Result<Self, Self::Error> {
        let chrom = record.reference_sequence_name().to_string();
        let pos = usize::from(
            record
                .variant_start()
                .ok_or(Self::Error::MissingVariantStart)?,
        ) as i32;

        let ref_allele = record.reference_bases().to_string();
        let alt_allele = record
            .alternate_bases()
            .as_ref()
            .iter()
            .next()
            .ok_or(Self::Error::MissingAlternateBases)?
            .to_string();

        Ok(Self {
            chrom,
            pos,
            ref_allele,
            alt_allele,
        })
    }
}

/// Helper type with mapping from sample name to `CallInfo`.
#[derive(Debug, Clone, Default)]
pub struct CallInfos {
    /// Mapping from sample name to `CallInfo`.
    pub call_infos: indexmap::IndexMap<String, CallInfo>,
}

/// Supporting code for `CallInfos`.
pub(crate) mod call_infos {
    use crate::common::GenotypeToStringError;

    /// Error type for `FromVcf` implementation.
    #[derive(thiserror::Error, Debug, Clone)]
    pub enum Error {
        #[error("Empty FORMAT/AD")]
        EmptyFormatAd,
        #[error("Problem parsing genotype: {0}")]
        GenotypeParsing(#[from] GenotypeToStringError),
    }
}

impl TryFromVcf for CallInfos {
    type Error = call_infos::Error;

    fn try_from_vcf(
        record: &vcf::variant::RecordBuf,
        header: &vcf::Header,
    ) -> Result<Self, Self::Error> {
        use vcf::variant::record::samples::keys::key;
        let mut result = indexmap::IndexMap::new();

        for (name, sample) in header.sample_names().iter().zip(record.samples().values()) {
            let genotype = if let Some(Some(
                vcf::variant::record_buf::samples::sample::value::Value::Genotype(gt),
            )) = sample.get(key::GENOTYPE)
            {
                Some(genotype_to_string(&gt)?)
            } else {
                None
            };
            let quality = if let Some(Some(
                vcf::variant::record_buf::samples::sample::value::Value::Integer(quality),
            )) = sample.get(key::CONDITIONAL_GENOTYPE_QUALITY)
            {
                Some(*quality as f32)
            } else {
                None
            };
            let dp = if let Some(Some(
                vcf::variant::record_buf::samples::sample::value::Value::Integer(dp),
            )) = sample.get(key::READ_DEPTH)
            {
                Some(*dp)
            } else {
                None
            };
            let ad =
                if let Some(Some(vcf::variant::record_buf::samples::sample::value::Value::Array(
                    vcf::variant::record_buf::samples::sample::value::Array::Integer(ad),
                ))) = sample.get(key::READ_DEPTHS)
                {
                    Some(ad[1].ok_or(Self::Error::EmptyFormatAd)?)
                } else {
                    None
                };
            let phase_set = if let Some(Some(
                vcf::variant::record_buf::samples::sample::value::Value::Integer(id),
            )) = sample.get(key::PHASE_SET)
            {
                Some(*id)
            } else {
                None
            };

            result.insert(
                name.clone(),
                CallInfo {
                    sample: name.clone(),
                    genotype,
                    gq: quality,
                    dp,
                    ad,
                    ps: phase_set,
                },
            );
        }

        Ok(CallInfos { call_infos: result })
    }
}

/// Helper type with `AnnField` records.
#[derive(Debug, Clone, Default)]
pub struct AnnFields {
    /// Mapping from sample name to `CallInfo`.
    pub ann_fields: Vec<mehari::annotate::seqvars::ann::AnnField>,
}

/// Supporting code for `AnnFields`.
pub(crate) mod ann_fields {
    /// Error type for `FromVcf` implementation.
    #[derive(thiserror::Error, Debug, Clone)]
    pub enum Error {
        #[error("Problem parsing INFO/ANN: {0}")]
        Parsing(String),
        #[error("Invalid type of INFO/ANN")]
        InvalidTypeInfoAnn,
    }
}

impl TryFromVcf for AnnFields {
    type Error = ann_fields::Error;

    fn try_from_vcf(
        record: &vcf::variant::RecordBuf,
        _header: &vcf::Header,
    ) -> Result<AnnFields, ann_fields::Error> {
        if let Some(Some(ann)) = record.info().get("ANN") {
            if let vcf::variant::record_buf::info::field::Value::Array(
                vcf::variant::record_buf::info::field::value::Array::String(ann),
            ) = ann
            {
                Ok(AnnFields {
                    ann_fields: ann
                        .iter()
                        .flatten()
                        .map(|s| s.parse::<mehari::annotate::seqvars::ann::AnnField>())
                        .collect::<Result<Vec<_>, _>>()
                        .map_err(|e| Self::Error::Parsing(format!("{}", e)))?,
                })
            } else {
                Err(ann_fields::Error::InvalidTypeInfoAnn)
            }
        } else {
            Ok(Default::default())
        }
    }
}

/// Sequence variant record.
#[derive(Debug, Clone, Default, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub struct VariantRecord {
    /// VCF-style variant.
    pub vcf_variant: VcfVariant,
    /// Call information.
    pub call_infos: indexmap::IndexMap<String, CallInfo>,
    /// Annotation fields.
    pub ann_fields: Vec<mehari::annotate::seqvars::ann::AnnField>,
    /// Population frequencies.
    pub population_frequencies: PopulationFrequencies,
}

/// Supporting code for `VariantRecord`.
pub(crate) mod variant_record {
    /// Error type for `FromVcf` implementation.
    #[derive(thiserror::Error, Debug, Clone)]
    pub enum Error {
        #[error("Problem with variant description: {0:?}")]
        VcfVariant(#[from] super::vcf_variant::Error),
        #[error("Problem with call information: {0:?}")]
        CallInfos(#[from] super::call_infos::Error),
        #[error("Problem with annotation fields: {0:?}")]
        AnnFields(#[from] super::ann_fields::Error),
        #[error("Problem with population frequencies: {0:?}")]
        PopulationFrequencies(#[from] super::population_frequencies::Error),
    }
}

impl TryFromVcf for VariantRecord {
    type Error = variant_record::Error;

    /// Convert from VCF record.
    ///
    /// # Arguments
    ///
    /// * `record` - VCF record.
    /// * `header` - VCF header.
    ///
    /// # Errors
    ///
    /// Returns an error if the record cannot be extracted.
    fn try_from_vcf(
        record: &vcf::variant::RecordBuf,
        header: &vcf::Header,
    ) -> Result<Self, Self::Error> {
        let vcf_variant = VcfVariant::try_from_vcf(record, header)?;
        let CallInfos { call_infos } = CallInfos::try_from_vcf(record, header)?;
        let AnnFields { ann_fields } = AnnFields::try_from_vcf(record, header)?;
        let population_frequencies = PopulationFrequencies::try_from_vcf(record, header)?;

        Ok(Self {
            vcf_variant,
            call_infos,
            ann_fields,
            population_frequencies,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[rstest::fixture]
    fn nuclear_frequencies() -> NuclearFrequencies {
        NuclearFrequencies {
            an: 100,
            hom: 10,
            het: 20,
            hemi: 5,
        }
    }

    #[rstest::fixture]
    fn mitochondrial_frequencies() -> MitochondrialFrequencies {
        MitochondrialFrequencies {
            an: 100,
            hom: 10,
            het: 20,
        }
    }

    #[rstest::fixture]
    fn inhouse_frequencies() -> InHouseFrequencies {
        InHouseFrequencies {
            hom: 10,
            het: 20,
            hemi: 5,
            an: 65,
        }
    }

    #[rstest::fixture]
    fn population_frequencies(
        nuclear_frequencies: NuclearFrequencies,
        mitochondrial_frequencies: MitochondrialFrequencies,
        inhouse_frequencies: InHouseFrequencies,
    ) -> PopulationFrequencies {
        PopulationFrequencies {
            gnomad_exomes: nuclear_frequencies.clone(),
            gnomad_genomes: nuclear_frequencies.clone(),
            gnomad_mtdna: mitochondrial_frequencies.clone(),
            helixmtdb: mitochondrial_frequencies.clone(),
            inhouse: inhouse_frequencies.clone(),
        }
    }

    #[rstest::rstest]
    fn test_nuclear_frequencies_an(nuclear_frequencies: NuclearFrequencies) {
        assert_eq!(nuclear_frequencies.an(), 100);
    }

    #[rstest::rstest]
    fn test_nuclear_frequencies_ac(nuclear_frequencies: NuclearFrequencies) {
        assert_eq!(nuclear_frequencies.hom(), 10);
        assert_eq!(nuclear_frequencies.het(), 20);
        assert_eq!(nuclear_frequencies.ac(), 45);
    }

    #[rstest::rstest]
    fn test_nuclear_frequencies_af(nuclear_frequencies: NuclearFrequencies) {
        assert_eq!(nuclear_frequencies.af(), 0.45);
    }

    #[rstest::rstest]
    fn test_nuclear_frequencies_hemi(nuclear_frequencies: NuclearFrequencies) {
        assert_eq!(nuclear_frequencies.hemi(), 5);
    }

    #[rstest::rstest]
    fn test_nuclear_frequencies_carriers(nuclear_frequencies: NuclearFrequencies) {
        assert_eq!(nuclear_frequencies.carriers(), 35);
    }

    #[rstest::rstest]
    fn test_mitochondrial_frequencies_an(mitochondrial_frequencies: MitochondrialFrequencies) {
        assert_eq!(mitochondrial_frequencies.an(), 100);
    }

    #[rstest::rstest]
    fn test_mitochondrial_frequencies_ac(mitochondrial_frequencies: MitochondrialFrequencies) {
        assert_eq!(mitochondrial_frequencies.hom(), 10);
        assert_eq!(mitochondrial_frequencies.het(), 20);
        assert_eq!(mitochondrial_frequencies.ac(), 30);
    }

    #[rstest::rstest]
    fn test_mitochondrial_frequencies_af(mitochondrial_frequencies: MitochondrialFrequencies) {
        assert_eq!(mitochondrial_frequencies.af(), 0.3);
    }

    #[rstest::rstest]
    fn test_mitochondrial_frequencies_carriers(
        mitochondrial_frequencies: MitochondrialFrequencies,
    ) {
        assert_eq!(mitochondrial_frequencies.carriers(), 30);
    }

    #[rstest::rstest]
    fn test_inhouse_frequencies_ac(inhouse_frequencies: InHouseFrequencies) {
        assert_eq!(inhouse_frequencies.hom(), 10);
        assert_eq!(inhouse_frequencies.het(), 20);
        assert_eq!(inhouse_frequencies.ac(), 45);
    }

    #[rstest::rstest]
    fn test_inhouse_frequencies_carriers(inhouse_frequencies: InHouseFrequencies) {
        assert_eq!(inhouse_frequencies.carriers(), 35);
    }

    #[rstest::rstest]
    fn test_population_frequencies(population_frequencies: PopulationFrequencies) {
        assert_eq!(population_frequencies.gnomad_exomes.an, 100);
        assert_eq!(population_frequencies.gnomad_genomes.an, 100);
        assert_eq!(population_frequencies.gnomad_mtdna.an, 100);
        assert_eq!(population_frequencies.helixmtdb.an, 100);
        assert_eq!(population_frequencies.inhouse.an, 65);
    }

    #[rstest::rstest]
    #[case::case_1_ingested_vcf("tests/seqvars/query/Case_1.ingested.vcf")]
    #[case::dragen_ingested_vcf("tests/seqvars/query/dragen.ingested.vcf")]
    pub fn sequence_variant_from_vcf(#[case] path_input: &str) -> Result<(), anyhow::Error> {
        use noodles::vcf::variant::RecordBuf;

        mehari::common::set_snapshot_suffix!("{}", path_input.split('/').last().unwrap());

        let mut vcf_reader = vcf::io::reader::Builder::default()
            .build_from_path(path_input)
            .unwrap();
        let header = vcf_reader.read_header()?;

        for record in vcf_reader.records() {
            let record = record?;
            let seqvar = super::VariantRecord::try_from_vcf(
                &RecordBuf::try_from_variant_record(&header, &record)?,
                &header,
            )?;

            insta::assert_yaml_snapshot!(&seqvar);
        }

        Ok(())
    }
}
