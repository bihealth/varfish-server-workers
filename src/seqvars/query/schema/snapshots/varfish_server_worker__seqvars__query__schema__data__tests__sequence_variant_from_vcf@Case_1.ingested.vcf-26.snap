---
source: src/seqvars/query/schema/data.rs
expression: "&seqvar"
---
vcf_variant:
  chrom: MT
  pos: 1824
  ref_allele: T
  alt_allele: C
call_infos:
  Case_1_father-N1-DNA1-WGS1:
    sample: Case_1_father-N1-DNA1-WGS1
    genotype: /1/1
    gq: 99
    dp: 2668
    ad: 2668
    ps: ~
  Case_1_index-N1-DNA1-WGS1:
    sample: Case_1_index-N1-DNA1-WGS1
    genotype: /0/0
    gq: 99
    dp: 2409
    ad: 0
    ps: ~
  Case_1_mother-N1-DNA1-WGS1:
    sample: Case_1_mother-N1-DNA1-WGS1
    genotype: /0/0
    gq: 99
    dp: 1752
    ad: 0
    ps: ~
ann_fields: []
population_frequencies:
  gnomad_exomes:
    an: 0
    hom: 0
    het: 0
    hemi: 0
  gnomad_genomes:
    an: 0
    hom: 0
    het: 0
    hemi: 0
  gnomad_mtdna:
    an: 0
    hom: 0
    het: 0
  helixmtdb:
    an: 0
    hom: 0
    het: 0
  inhouse:
    an: 0
    hom: 0
    het: 0
    hemi: 0
